File: 01_array/java/AddBinary.java
/*
Given two binary strings, return their sum (also a binary string).

For example,
a = "11"
b = "1"
Return "100".
*/


/* My Solution follows the Two Pointer Strategy */

public class Solution {
    public String addBinary(String a, String b) {
        if(a == null || b == null) {
            return null;
        }
        
        int m = a.length();
        int n = b.length();
        
        if(m == 0 || n == 0) {
            return null;
        }
        
        int carry = 0;
        String string = "";
        
        while ( m>0 || n>0 ) {
            
            int sa = 0;
            int sb = 0;
            
            if (m>0) {
                sa = a.charAt(m-1) - '0';
            }
            
            if (n>0) {
                sb = b.charAt(n-1) - '0';
            }
            
            int temp = carry + sa + sb;
            string = temp % 2 + string;
            carry = temp/2;
            
            m--;
            n--;
        }
        
        return (carry==0)? string: "1" + string;
    }
}

File: math/add_binary/add_binary.cpp
/*
67. Add Binary
https://leetcode.com/problems/add-binary/

Given two binary strings, return their sum 
(also a binary string).

For example,
a = "11"
b = "1"
Return "100".
*/


#include <string>
#include <iostream>

using namespace std;

class Solution {
public:
    string addBinary(string a, string b) {
        if (a.empty() || b.empty()) {
            return "";
        }
        
        int a_len = a.length();
        int b_len = b.length();
        
        int carry = 0;
        string result("");
        
        // adding from the end
        while (a_len>0 || b_len >0) {
            int sa = 0;
            int sb = 0;
            
            if (a_len>0) {
                sa = a[a_len-1] - '0';
            }
            
            if (b_len >0) {
                sb = b[b_len-1] - '0';
            }
            
            int temp = carry + sa + sb;
            result = to_string(temp%2) + result;
            carry = temp/2;
            
            a_len--;
            b_len--;
        }
        
        //cout << "result value is: " << result << endl;
        //cout << "carry value is: " << carry << endl;

        return (carry==0) ? result : "1" + result;
    }
};


int main() {
	string a("1");
	string b("1");

	Solution s;
	cout << "1 + 1 = " << s.addBinary(a, b) << endl;

	return 0;
}